### PR TITLE
Support for software list & update on child devices

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

Implement `software_list` support as described by https://github.com/thin-edge/thin-edge.io/pull/2235

- [x] Use `te` MQTT scheme
- [x] Use commands instead of requests / responses for software list
- [x] Update the agent to publish `software-list` capability on start
- [x] Update the agent to respond to `software-list` requests
- [x] Update the c8y mapper to forward software list request to child devices when their agent starts.
- [x] Update the c8y mapper to request a software list request after a software-update operation (successful or not)

Implement `software_update` support as described by https://github.com/thin-edge/thin-edge.io/pull/2235

- [x] Use commands instead of requests / responses for software update
- [x] Update the agent to publish `software-update` capability on start
- [x] Update the agent to respond to `software-update` requests
- [x] Update the c8y mapper to to forward software update requests
- [ ] Publish the list of supported software types along the capability message. See https://github.com/thin-edge/thin-edge.io/issues/1352

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [inline crates/core/tedge_api/src/messages.rs:93]: TODO: be consistent with `RestartCommand` constructors.

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [inline crates/core/tedge_api/src/messages.rs:172]: TODO: find a way (a `trait` extension) to avoid repeating all these methods (`status()`, `topic_id()`, `channel()` ...) for all the command types.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (Bravo555) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1305]: note: It's not obvious what kind of validation does this function perform, so I'd like to see a doc-comment here, as this isn't a new function, it's not a blocker.

**Review by Bravo555 — APPROVED** [review]

**Review by albinsuresh — COMMENTED** [review]
LGTM. Happy to approve after a test run once the merge conflicts are resolved.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: Indeed, not a good name. Fixed with https://github.com/thin-edge/thin-edge.io/pull/2307/commits/4ef53919ee13e959d61a8129f8688865e9821036

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:1321]: Indeed, it would be good to have operation specific modules or even actor extensions. However, I would prefer to address that later once both software list and software update fully ported to support child devices.

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1220]: No, at least not with the current implementation which sends a new `114` message each time a new capability message is received.

On start, `tedge-agent` sends two capabilities, one for restart and another for software-list.
Hence, c8y mapper sends two `114` messages, one adding support for `c8y_Restart` and a second one adding support for `c8y_SoftwareUpdate` (see the following assert).

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:1226]: I don't get your comment. This type is here to *hold* an actual operation request, not to declare an operation capability. Am I missing something?

Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:1226]: Oops sorry. I thought we had an enum with all the supported operation variants listed, but unfortunately no. We probably should still have one in the `smartrest` module to avoid passing operation names as `&str`s like this.

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1220]: Aah sorry, I was confused to see the `114` message with `c8y_Restart` imediately after sending the `cmd/software_list` metadata message in the very last step. Keeping this specific assertion for `c8y_Restart` one step above, right after the `cmd/restart` message is sent, would have avoided that confusion.

**Review by albinsuresh — COMMENTED** [review]
Happy to approve once the system test failures are sorted.

**Comment by didier-wenzek** [comment]
The [failing system test is related to self-update](https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/self-update/tedge_self_update.robot): thin-edge used to update thin-edge on the device.

So issue is that this test doesn't update `tedge-agent` but only `tedge-mapper` (and other tedge executables unrelated to the issue). But the new `tedge-mapper` is not compatible with the old `tedge-agent` on how a software-list is requested. Hence, the software list is is not updated on c8y, failing the test.

**Comment by reubenmiller** [comment]
> The [failing system test is related to self-update](https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/self-update/tedge_self_update.robot): thin-edge used to update thin-edge on the device.
> 
> So issue is that this test doesn't update `tedge-agent` but only `tedge-mapper` (and other tedge executables unrelated to the issue). But the new `tedge-mapper` is not compatible with the old `tedge-agent` on how a software-list is requested. Hence, the software list is is not updated on c8y, failing the test.

So this is the first problem that has popped up due to the fact that the `tedge-agent` service is not restarted during the installation process, so the old instance continues running after the new version has been installed. We will have to solve this in the workflow, otherwise it will lead to very unexpected results for users.

**Comment by didier-wenzek** [comment]
> > The [failing system test is related to self-update](https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/self-update/tedge_self_update.robot): thin-edge used to update thin-edge on the device.
> > So issue is that this test doesn't update `tedge-agent` but only `tedge-mapper` (and other tedge executables unrelated to the issue). But the new `tedge-mapper` is not compatible with the old `tedge-agent` on how a software-list is requested. Hence, the software list is is not updated on c8y, failing the test.
> 
> So this is the first problem that has popped up due to the fact that the `tedge-agent` service is not restarted during the installation process, so the old instance continues running after the new version has been installed. We will have to solve this in the workflow, otherwise it will lead to very unexpected results for users.

I was happy with an idea for a workaround : using the new all-in-one `tedge` executable to turn it into a proper version of `tedge-agent` using a link ... but this is not working for reasons I fail to understand for now. Despite the agent reports the expected version on the device, the versions are not updated on c8y.

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs]: Addresses by https://github.com/thin-edge/thin-edge.io/pull/2307/commits/ecc2a84b7883650a1dadb5bce23e1b127db18872

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs]: Addresses by https://github.com/thin-edge/thin-edge.io/pull/2307/commits/ecc2a84b7883650a1dadb5bce23e1b127db18872

**Review by didier-wenzek — COMMENTED** [review]
Ready for review

File: crates/core/tedge_agent/src/tedge_operation_converter/actor.rs
Comment (rina23q) [inline crates/core/tedge_agent/src/tedge_operation_converter/actor.rs:108]: To keep the consistency, use `_` instead of `-`. We use underscore for the predefined command types: https://thin-edge.github.io/thin-edge.io/next/references/mqtt-api/#command-examples

```suggestion
                Err(err) => error!("Incorrect software_list request payload: {err}"),
```

**Review by rina23q — COMMENTED** [review]
I just found you use `-` in another PR for command type names, so I jumped in here to be "hyphen police" ;)

File: crates/core/tedge_agent/src/tedge_operation_converter/actor.rs
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/tedge_operation_converter/actor.rs]: Fixed by https://github.com/thin-edge/thin-edge.io/pull/2307/commits/3e8d8f8615f5cf4e012601bd23d68634f6e7384a

File: crates/core/tedge_api/src/lib.rs
Comment (albinsuresh) [inline crates/core/tedge_api/src/lib.rs:929]: Remove that test? Looks like an irrelevant test as software update response doesn't carry the software list anymore.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:1174]: Actually there is one: `CumulocitySupportedOperations`.

**Review by albinsuresh — COMMENTED** [review]
The core changes look good. Just some refactoring suggestions.

File: tests/RobotFramework/tests/cumulocity/software_management/software-update-child.robot
Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/software_management/software-update-child.robot]: Fixed with https://github.com/thin-edge/thin-edge.io/pull/2307/commits/ddbe805f7ffc9bdac17ef945af5d608b2a2b9c79

File: tests/RobotFramework/tests/cumulocity/software_management/software-update-child.robot
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/software_management/software-update-child.robot:31]: Yes we still have some conceptual work to do here as a follow up, however I wouldn't look at the exact solution until we consolidate the number of services/processes.

Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/software_management/software-update-child.robot:31]: However running thin-edge.io in a container is easier to change this as you can set an environment variable to affect the behaviour of each component. systemd does not support inheriting environment variables.

File: crates/core/tedge_agent/src/software_manager/actor.rs
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/software_manager/actor.rs]: Addressed by https://github.com/thin-edge/thin-edge.io/pull/2307/commits/9555293c26a8ed5670bd23d693d66911ab9aa4d2

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs:29]: Having the schema attached to a message looks upside down.

I would prefer to address this in a more generic manner as proposed here https://github.com/thin-edge/thin-edge.io/issues/2357

**Review by reubenmiller — APPROVED** [review]
Approved. Great effort!

File: crates/extensions/c8y_mapper_ext/src/compatibility_adapter.rs
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/compatibility_adapter.rs:166]: I'd add some unit test coverage, **if there's time**, especially for the `convert_to_old_agent_request` function which has some deep nested logic.

**Review by albinsuresh — APPROVED** [review]
LGTM

File: tests/RobotFramework/tests/cumulocity/self-update/tedge_self_update.robot
Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/self-update/tedge_self_update.robot:38]: Yes. This is the case since the work on self-update. Restarting the agent during a software update operation leads to a fail operation status.

This is something we will need to fix, sure.

**Pull request merged**
